service/history/workflowResetter.go

```diff
@@ -138,7 +145,52 @@ func (r *workflowResetterImpl) resetWorkflow(
 			return err
 		}
 		resetWorkflowVersion = currentMutableState.GetCurrentVersion()
-		currentWorkflowTerminated = true
+
+		currentWorkflowMutation, currentWorkflowEventsSeq, err = currentMutableState.CloseTransactionAsMutation(
+			r.shard.GetTimeSource().Now(),
+			workflow.TransactionPolicyActive,
+		)
+		if err != nil {
+			return err
+		}
+		currentWorkflowMutation.ExecutionInfo.ExecutionStats = &persistencespb.ExecutionStats{
+			HistorySize: currentWorkflow.getContext().GetHistorySize(),
+		}
+
+		reapplyEventsFn = func(ctx context.Context, resetMutableState workflow.MutableState) error {
+			if err := r.reapplyContinueAsNewWorkflowEvents(
+				ctx,
+				resetMutableState,
+				namespaceID,
+				workflowID,
+				baseRunID,
+				baseBranchToken,
+				baseRebuildLastEventID+1,
+				baseNextEventID,
+			); err != nil {
+				return err
+			}
+
+			for _, event := range currentWorkflowEventsSeq {
+				if err := r.reapplyEvents(resetMutableState, event.Events); err != nil {
```

Comment:
what if we fail here? the workflow would already been reseted, right?

Reply:
is it easer to flush the buffered events (with force close current in-flight workflow task), then do normal reset?

Reply:
>what if we fail here? the workflow would already been reseted, right?

no, changes are persisted to DB at line 228 (after the change.)

>is it easer to flush the buffered events (with force close current in-flight workflow task), then do normal reset?

are you suggesting split the operation in 2?